src/clients/Client.cc

```diff
@@ -230,6 +219,16 @@ void
 Client::completeForwarding()
 {
     debugs(11,5, "completing forwarding for "  << fwd);
+
+    const char *receivedWholeReply = receivedWholeVirginReply ? "complete virgin reply" : nullptr;
+#if USE_ADAPTATION
+    if (startedAdaptation)
```

Comment:
> > Can we guarantee that completeForwarding() is never called until all deferred adapted bytes are written to Store?

> I think that the code assumes that.

I suspect you are right regarding the existence of that assumption in the current code (but I want this RP to go further than confirming existing code _assumptions_).

> calls cleanAdaptation() ... removing adaptedBodySource ... thus assuming that all adaptedBodySource bytes are consumed

FWIW, I do not think that a call to cleanAdaptation() or removing adaptedBodySource implies that all adaptedBodySource bytes were consumed. cleanAdaptation() is simply disassociating Client from BodyPipe. That cleanup happens on both "consumed everything" and "did not consume something" cases. This paragraph is _not_ meant to invalidate your _conclusion_ regarding the "completeForwarding() is never called until all deferred adapted bytes are written to Store" assumption. Again, I think that conclusion is correct.

Reply:
receivedWholeAdaptedReply means that handleAdaptedBodyProductionEnded() was called (while the Store entry was not aborted). Current PR code implies that the lack of handleAdaptedBodyProductionEnded() call (i.e. false receivedWholeAdaptedReply) means that we have _not_ received the entire adapted response.

I am concerned about a possible race condition between handleAdaptedBodyProductionEnded() and the final resumeBodyStorage() callbacks. Only the former sets receivedWholeAdaptedReply, but both trigger endAdaptedBodyConsumption() ... handleAdaptationCompleted() ... completeForwarding() path. We may be missing a successful termination case here -- not marking complete Store entries as such because we end transaction too early.

N.B. IIRC, Server code still we has bugs/workarounds related to [similar race conditions](https://bugs.squid-cache.org/show_bug.cgi?id=5187#c6) associated with noteBodyProductionEnded() callbacks, so this is not a purely theoretical concern. See XXX in ClientHttpRequest::noteMoreBodyDataAvailable(). This PR might even provide a blueprint for fixing that code!

N.B. adaptedBodySource->exhausted() does _not_ imply that the whole adapted message body was received. The method also returns true when a _partial_ adapted message body was received (and consumed) after adapted body delivery was aborted. Callers must respect that method semantics.

There are many ways to address my concern. How about starting with the following sketch?

* Declare that endAdaptedBodyConsumption() is meant for handling successful adaptation termination cases only. Non-successful cases should abort/end the current transaction without calling endAdaptedBodyConsumption(). We might even rename that method at some point. The current PR code may already comply with this declaration, but I have not checked.
* Require a prior or current handleAdaptedBodyProductionEnded() call to endAdaptedBodyConsumption(). Thus, do not just use adaptedBodySource->exhausted() condition in resumeBodyStorage() to endAdaptedBodyConsumption(). resumeBodyStorage() should _also_ use receivedWholeAdaptedReply precondition for calling endAdaptedBodyConsumption(). resumeBodyStorage() should also terminate the transaction if adaptedBodySource->exhausted() is true and noteBodyProducerAborted() has been called for the adapted response; we must not wait for handleAdaptedBodyProductionEnded() that will never come.

P.S. This suggestion is still working within the current PR model that attempts to use "got successful-end-of-message and have an empty message body buffer" condition (instead of comparing received/written body byte counts) as markStoredReplyAsWhole() precondition.